Ability to Specify Delay in PML

laurmurclar opened the pull request:
#154

laurmurclar committed:
Add ability to specify delays to parser

laurmurclar committed:
Better test files

laurmurclar commented on panacea/src/pml_parser.yrl:
```diff
@@ -169,6 +191,45 @@ operator -> '>=' : ">=".
 
 Erlang code.
 
+validate_time({TimeType, [{line, Line}], Val}) ->
```
Anyone know of a nicer way to write this function? My Erlang is in the very early stages of learning so I'm not exactly sure what the options are..

laurmurclar replied:
I guess even a function to wrap the `return_error` with the values like `time_out_of_range_error(TimeType, MaxVal, ActualVal)` would be better than what is there now.

houli commented on panacea/src/pml_parser.yrl:
```diff
@@ -127,10 +134,25 @@ action_attribute ->
     expression : '$1'.
 
+time_expr ->
+    years '{' number '}'           : construct('$1', [], value_of('$3')).
```
Think `number` is too general here because it can have floats and scientific notation and what not. We should have like a positive integer token

c-brenn replied:
Good point. Floats etc would prob cause the `list_to_integer` call to error as well.

laurmurclar replied:
Sounds good

houli replied:
The only thing that might cause a problem is introducing ambiguity in the parser. God speed.

c-brenn replied:
If you put the `integer` token above the `number` one in the lexer then all is good (I think).

houli commented on panacea/src/pml_parser.yrl:
```diff
@@ -169,6 +191,45 @@ operator -> '>=' : ">=".
+  NewVal = list_to_integer(Val),
+  case TimeType of
+    years when NewVal >= 100 ->
+      return_error(Line, "years cannot be more than 99 (was " ++ integer_to_list(NewVal) ++ ")");
```
In all of these you can just refer to Val because you're converting and then converting back?

laurmurclar replied:
good spot 👍

c-brenn commented on panacea/src/pml_parser.yrl:
```diff
@@ -127,10 +134,25 @@ action_attribute ->
 requires_expr ->
     expression : '$1'.
 
+time_expr ->
```
Do you think this should return a `construct` instead of just an integer?

I'm thinking about taking an AST and producing PML again, with this approach I think we lose everything inside a `time` block.

Something like:

`construct('$1', [{minutes, get_time('$3')}], validate_times('$3'))`

Which would produce: `{:time, [minutes: N], [list of inner times]}`

laurmurclar replied:
Never considered recreating it actually, good point. Seems like a good solution

laurmurclar committed:
Suggested changes from PR

laurmurclar committed:
Remove time summing from parser

laurmurclar commented:
So writing the elixir module is trickier than I anticipated. I think it's worth just adding the Erlang version for now to make sure it's included in this iteration. Can abstract out if I have time next week, but I definitely won't this weekend (my FYP demo is Monday)

laurmurclar commented:
Can add some tests tonight but that's about all I'll have a chance to do this weekend.

c-brenn commented:
Looks good. It's much simper now so I think having it all in Erlang is perfect. If we have to do anything more complicated with time (unlikely) it might be worth reconsidering the elixir module.

I have some free time today so I can finish off the tests and AST -> PML bit if you'd like.

laurmurclar commented:
I'm happy I'll get the tests done but I'm not sure I'll have time to do the AST -> PML bit today. I can work on it tomorrow if you don't want to do it?

c-brenn commented:
👍 work away on the tests so. The ast bit should only be a few small changes for me, I'll add that in after.

c-brenn commented on panacea/src/pml_lexer.xrl:
```diff
@@ -1,16 +1,22 @@
+IDENT   = [a-zA-Z0-9_]+
+WS      = [\s\t\n\r]+
+NATURAL = [0-9]+
+NUMBER  = ({NATURAL}(\.{NATURAL})?|\.{NATURAL})([eE][+-]?{NATURAL})?
```
What do you think about changing `NUMBER` to something like `FLOAT`?
At the moment it still matches natural numbers as well as floats - seeing as we now have `NATURAL` we could remove that bit of the regex:

```
FLOAT  = ({NATURAL})?\.{NATURAL}([eE][+-]?{NATURAL})?
```

something like the above.

c-brenn commented on panacea/src/pml_lexer.xrl:
```diff
@@ -50,6 +56,7 @@ executable    : {token, {executable, TokenLine}}.
 {WS}          : skip_token.
 {STRING}      : {token, {string, TokenLine, TokenChars}}.
+{NATURAL}     : {token, {natural, TokenLine, TokenChars}}.
 {NUMBER}      : {token, {number, TokenLine, TokenChars}}.
```
would involve changing this to `float` as well - in the lexer and parser

c-brenn committed:
changes number regexes in lexer

We split numbers into integers and floats so that we could restrict
times to integers. This changes the names of the tokens to `integer` and
`float` - used to be `natural` and `number`.

The `number` (now `float`) regex used to match integers and floats, it
now only matches floats.

c-brenn committed:
adds initial tests for time parsing

c-brenn committed:
Adds time to Ast.ex

This change allows Ast.ex to deal with times in the PML AST. To
facilitate this, it changes `:requires` into a multi-line construct.
`:requires` is different to other multi-line constructs in that it has a
single child node instead of a list of children. To deal with this, I've
added another `do_unquote` clause that wraps the child node in a list.

c-brenn committed:
changes how the parser handles numbers

This change was going to just add a test to ensure the parser properly
deals with non-integer time values. With the current lexer and parser
implementation, any error message containing a number would look like
nonsense - because yecc keeps the tokens as a charlist.

We expect a message like:

`syntax error before: 4.5`

But we'd actually get something like:

`syntax error before: [49,52,34]`

This changes how the lexer handles numbers to improve these error
messages. The lexer turns the character list representing a number into
an actual number (int or float). This allows the parser to print a nice
message. For simplicity, we've kept all expressions in the AST as
strings, so I've also changed the parser to turn the numbers back into
charlists when it's building the AST.

c-brenn committed:
adds e2e tests for delays

laurmurclar commented:
Good job @c-brenn. Nice change with the floats 👍
